william-wtr92/instamint
# feat(2FA) : 🚀 users can enable/disable two-factor authentication

# [ITM-21](https://project-william.atlassian.net/browse/ITM-21) As a MINTER, I can add or remove a 2FA authentication

This feature allows users to enable/disable two-factor authentication to add an extra layer of security to their accounts.

## Proposed Changes

These changes concern the implementation of two-factor authentication using `otplib` library, enabling users to activate/deactivate 2FA and increasing account security by asking a 2FA code for more "risky" tasks.

- Added :
  - View : Two factor authentication sign-in modal
  - View : Two factor authentication enable/disable modal
  - Server : API routes to handle two-factor authentication
  - Server : Implementation of `otplib` actions for all types of OTP

- Updated : 
    - Client : Services type -> now accepts type for parameter and for returned value
    - Server : base sign-in API routes now handle two-factor authentication

## Type of Change

What type of change are you making? (Check all that apply)


## How Has This Been Tested?

Explain the tests you conducted to validate your changes. Provide clear instructions so that the code review can
replicate your tests and confirm the results.

- To test this ticket, follow these instructions : 
    - Run at root `pnpm install`
    - Update your `env variables` to `.env.example`
    - Do migrations with `pnpm run kn:latest`
    - Run the client and server side : in two different terminal, run `pnpm run dev:webapp` and `pnpm run dev:business`
    - Sign-in into the application 
    - Go into this page `http://localhost:3000/profile/settings/security`
    - Follow the instructions in the modal used to activate two-factor authentication
    - Once it is configured on your chosen 2FA authentication app (like `Google Authenticator`), sign out to try it out
    - A two-factor authentication code will be needed to log-in (or your backup codes that can be used only one time each), otherwise you won't be able to sign-in back

## Checklist

Before submitting your PR, please check the appropriate boxes to confirm that you have taken the following actions:


## Screenshots

https://github.com/william-wtr92/instamint/assets/90528753/234af012-cde1-4301-bd64-eb47a8c12c2d


[ITM-21]: https://project-william.atlassian.net/browse/ITM-21?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
